x/nft/msg_test.go

```diff
@@ -0,0 +1,94 @@
+package nft
+
+import (
+	"testing"
+
+	"github.com/iov-one/weave/x"
+	. "github.com/smartystreets/goconvey/convey"
+)
+
+func TestApprovalMsg(t *testing.T) {
+	Convey("Test add/remove approvals msg validation", t, func() {
+		var helper x.TestHelpers
+		_, validKey := helper.MakeKey()
+		Convey("Test add approvals", func() {
+			Convey("Happy flow", func() {
+				msg := AddApprovalMsg{
+					Id:      []byte("asdf"),
+					Address: validKey.Address(),
+					Action:  ActionUpdateDetails,
+				}
+				Convey("Positive count", func() {
+					msg.Options = ApprovalOptions{Count: 1}
+					So(msg.Validate(), ShouldBeNil)
+				})
+				Convey("Unlimited count", func() {
+					msg.Options = ApprovalOptions{Count: UnlimitedCount}
+					So(msg.Validate(), ShouldBeNil)
+				})
+			})
+
+			Convey("Testing various errors", func() {
+				msg := AddApprovalMsg{
+					Id:      []byte("asdf"),
+					Address: validKey.Address(),
+					Action:  ActionUpdateDetails,
+				}
+
+				Convey("Invalid action", func() {
+					msg.Action = "asd"
+					So(msg.Validate(), ShouldNotBeNil)
+				})
+
+				Convey("Invalid id", func() {
+					msg.Id = []byte("asd")
```

Comment:
req: assuming tests are executed top to bottom, you would need to restore  `msg. Action` to a valid value before testing for invalid ID. Otherwise you can not really tell if invalid ID caused the validation failure.

Reply:
@alpe you have to go through the link I sent to see how tests are being executed in goconvey.

Parent setup is executed every time for each test, so there is no need in extra code to clean up!

Reply:
thanks for clarifying this!